src/ripple/rpc/handlers/DownloadShard.cpp

```diff
@@ -0,0 +1,320 @@
+    if (context.role != Role::ADMIN)
+        return rpcError(rpcNO_PERMISSION);
+
+    // The shard store must be configured
+    if (!context.app.getShardStore())
+        return rpcError(rpcNOT_ENABLED);
+
+    // Deny request if already downloading
+    if (context.app.getShardStore()->getNumPreShard())
+        return rpcError(rpcTOO_BUSY);
+
+    if (!context.params.isMember(jss::shards))
+        return RPC::missing_field_error(jss::shards);
+    if (!context.params[jss::shards].isArray() ||
+        context.params[jss::shards].size() == 0)
+    {
+        return RPC::make_error(
+            rpcINVALID_PARAMS,
+            RPC::expected_field_message(
+                jss::shards,
+                "an array"));
+    }
+
+    auto dCtx {std::make_shared<DownloadContext>()};
+    dCtx->shardStore = context.app.getShardStore();
+    dCtx->j = context.app.journal("DownloadShard");
+
+    // Validate shards
+    static const std::string ext {".tar.lz4"};
+    for (auto& it : context.params[jss::shards])
+    {
+        // Validate the index
+        if (!it.isMember(jss::index))
+            return RPC::missing_field_error(jss::index);
+        auto& jv {it[jss::index]};
+        if (!(jv.isUInt() || (jv.isInt() && jv.asInt() >= 0)))
+        {
+            return RPC::make_error(rpcINVALID_PARAMS,
+                RPC::expected_field_message(jss::index,
+                    "unsigned integer"));
+        }
+
+        // Validate the URL
+        if (!it.isMember(jss::url))
+            return RPC::missing_field_error(jss::url);
+        parsedURL url;
+        if (!parseUrl(url, it[jss::url].asString()) ||
+            url.scheme != "https" ||
+            url.domain.empty() ||
+            url.path.empty())
+        {
+            return RPC::invalid_field_error(jss::url);
+        }
+
+        // URL must point to tar archive using lz4 compression
+        auto archiveName {url.path.substr(url.path.find_last_of("/\\") + 1)};
+        if (archiveName.empty() || archiveName.size() <= ext.size())
+        {
+            return RPC::make_error(
+                rpcINVALID_PARAMS,
+                RPC::expected_field_message(
+                    jss::url,
+                    "invalid archive name"));
+        }
+        if (!boost::iends_with(archiveName, ext))
+        {
+            return RPC::make_error(
+                rpcINVALID_PARAMS,
+                RPC::expected_field_message(
+                    jss::url,
+                    "invalid archive extension"));
+        }
+
+        dCtx->shards.emplace(jv.asUInt(), std::move(url));
+    }
+
+    // Prepare the shard store
+    RangeSet<std::uint32_t> indexes;
+    for (auto const& shard : dCtx->shards)
+        indexes.add(shard.first);
+    indexes = dCtx->shardStore->prepareShard(indexes);
+    if (indexes.empty())
+        return RPC::make_param_error("invalid shard ids");
+
+    // Remove unsuitable shards
+    for (auto it = dCtx->shards.begin(); it != dCtx->shards.end();)
```

Comment:
What might cause a shard to be "unsuitable" - would it be if it's already in the shardstore and/or is in progress via peernet fetch? Just wondering if there is any reason to report this info back to the user...like "we already have shards x,y" or "we're already working on z".

Reply:
Yes, those cases make shard ids unsuitable. Keep in mind this is an admin command and the ids that will be downloaded are reported. I can add the omitted ids to the report but not sure if there is sufficient reason to. What do you think?

Reply:
at a minimum, I guess we should make a note to document the behavior - that is, if you ask for x,y,z only only get z back, then x and y were skipped for <potential reasons> and see the logs for more info (?). I think the documentation approach is reasonable for this fairly specialized command.